### PR TITLE
SR-3131: Adjust choice of decimal vs. exponential format

### PR DESCRIPTION
**Pull request by tbkka**
For each floating-point type, there is a range of integers which
can be exactly represented in that type.  Adjust the formatting
logic so that we use decimal format for integers within this
range, exponential format for numbers outside of this range.

For example, Double has a 53-bit significand so can exactly
represent every integer from `-(2^53)...(2^53)`.  With this
change, we now use decimal format for these integers and
exponential format for values outside of this range.  This is
a relatively small change from the previous logic -- we've
basically just moved the cutoff from 10^15 to 2^53 (≈ 10^17).

The decision for using exponential format for small numbers is
not changed.

This includes edits to the test suite to adapt to the change in
behavior.

Resolves #42728.

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please smoke test

File: stdlib/public/runtime/SwiftDtoa.cpp [hunk 1]
```diff
@@ -1166,13 +1170,16 @@ size_t swift_format_float80(long double d, char *dest, size_t length)
...
     int decimalExponent;
     int8_t digits[21];
     int digitCount =
         swift_decompose_float80(d, digits, sizeof(digits), &decimalExponent);
-    if (decimalExponent < -3 || decimalExponent > 18) {
+    // People use long double to model integers <= 2^64, so we use that
+    // as a cutoff for decimal vs. exponential format.
+    // The constant is written out in full here since it can't be
+    // expressed as a 64-bit integer.
+    if (decimalExponent < -3 || fabsl(d) > 18446744073709551616.0L) {
```
Comment (stephentyrone) [inline stdlib/public/runtime/SwiftDtoa.cpp]: Let's write this as `0x1.0p64L`?

**Comment by tbkka** [comment]
@swift-ci Please test
